File: src/shared/eval/families/aggression.ts
import type { Color } from "chessops/types";
import { opposite } from "chessops/util";

import { CLASSICAL_VALUES } from "../../chess";
import { featureId } from "../features";
import type { FeatureVector } from "../vector";
import type { EvalContext } from "./context";
import { relativeRank } from "./masks";

const OPPONENT_MOBILITY = featureId("opponentMobility");
const PUSH_DEPTH = featureId("pushDepth");
const OFFERED_MATERIAL = featureId("offeredMaterial");

export const SLOTS = [OPPONENT_MOBILITY, PUSH_DEPTH, OFFERED_MATERIAL];

// Squares a side's pieces can move to, pawn pushes aside. Real legal-move generation would mean
// flipping the side to move, which is not even a legal position when the other king is in check,
// so this counts attacked squares instead — the same approximation mobility makes.
function reachable({ context, color }: { context: EvalContext; color: Color }): number {
	const own = context.position.board[color];
	let total = 0;

	for (const { piece, reach } of context.reach) {
		if (piece.color === color) total += reach.diff(own).size();
	}

	return total;
}

// How far into enemy territory a side has pushed, counting only what has crossed the halfway
// line. Deliberately not the `advancement` feature, which counts every pawn's rank and would give
// the tuner two ways to say a related thing.
function depth({ context, color }: { context: EvalContext; color: Color }): number {
	let total = 0;

	for (const square of context.position.board[color]) {
		total += Math.max(0, relativeRank({ color, square }) - 3);
	}

	return total;
}

// Material a side is leaving to be taken, counted once per way it can be captured — a piece three
// pieces attack is offered three times over. `generous` wants this high, everything else low.
function offered({ context, color }: { context: EvalContext; color: Color }): number {
	const { board } = context.position;
	const enemy = opposite(color);
	let total = 0;

	for (const { piece, reach } of context.reach) {
		if (piece.color !== enemy) continue;

		for (const target of reach.intersect(board[color])) {
			total += CLASSICAL_VALUES[board.getRole(target) ?? "king"];
		}
	}

	return total;
}

// Pressure, in the three flavours the paper found: taking the opponent's moves away, pushing
// everything forward regardless of safety, and handing out material on purpose.
export function extractAggression({
	context,
	features,
}: {
	context: EvalContext;
	features: FeatureVector;
}): void {
	// Three separate walks, so three separate gates — a bot naming one of them has no use for the
	// other two.
	if (context.weighs(OPPONENT_MOBILITY)) {
		features[OPPONENT_MOBILITY] = reachable({ context, color: context.them });
	}

	if (context.weighs(PUSH_DEPTH)) {
		features[PUSH_DEPTH] =
			depth({ context, color: context.us }) - depth({ context, color: context.them });
	}

	if (context.weighs(OFFERED_MATERIAL)) {
		features[OFFERED_MATERIAL] =
			offered({ context, color: context.us }) - offered({ context, color: context.them });
	}
}

File: engine/src/eval/aggression.c
#include "bitboard.h"
#include "eval.h"
#include "extractors.h"
#include "masks.h"
#include "position.h"
#include "values.h"

// Squares a side's men attack that its own men do not stand on, pawns and king included — the
// stand-in for their legal moves, which would need the other side to move.
static int reachable(EvalContext *ctx, Color color) {
	Bitboard own = ctx->pos->colors[color];
	int total = 0;
	for (Bitboard men = own; men != 0;) {
		total += bb_count(ctx->reach[bb_pop(&men)] & ~own);
	}
	return total;
}

// How far past the halfway line a side's men stand, counted from the line.
static int push_depth(EvalContext *ctx, Color color) {
	int total = 0;
	for (Bitboard men = ctx->pos->colors[color]; men != 0;) {
		int past = relative_rank(color, bb_pop(&men)) - 3;
		total += past > 0 ? past : 0;
	}
	return total;
}

// Material `color` leaves to be taken, once per enemy man that could take it.
static int offered(EvalContext *ctx, Color color) {
	const Position *pos = ctx->pos;
	int total = 0;
	for (Bitboard enemies = pos->colors[opposite(color)]; enemies != 0;) {
		for (Bitboard targets = ctx->reach[bb_pop(&enemies)] & pos->colors[color]; targets != 0;) {
			total += classical_value(piece_role(pos->board[bb_pop(&targets)]));
		}
	}
	return total;
}

// Taking the opponent's moves away — a raw count of theirs, nothing to subtract.
float extract_opponent_mobility(EvalContext *ctx) {
	eval_walk(ctx);
	return (float)reachable(ctx, ctx->them);
}

// Pushing everything forward: the one aggression feature that needs no walk.
float extract_push_depth(EvalContext *ctx) { return side_difference(ctx, push_depth); }

// Handing out material on purpose.
float extract_offered_material(EvalContext *ctx) {
	eval_walk(ctx);
	return side_difference(ctx, offered);
}
